src/services/blocks/BlocksService.ts

```diff
@@ -450,84 +474,122 @@ export class BlocksService extends AbstractService {
+			(this.minCalcFeeRuntime && specVersion < this.minCalcFeeRuntime) ||
 			typeof api.query.transactionPayment?.nextFeeMultiplier?.at !== 'function'
 		) {
-			// We do not have the necessary materials to build calcFee, so we just give a dummy function
-			// that aligns with the expected API of calcFee.
-			calcFee = null;
-
-			const version = await api.rpc.state.getRuntimeVersion(parentHash);
-			[specVersion, specName] = [
-				version.specName.toString(),
-				version.specVersion.toNumber(),
-			];
-		} else {
-			const coefficients = api.consts.transactionPayment.weightToFee.map(
-				(c) => {
-					return {
-						// Anything that could overflow Number.MAX_SAFE_INTEGER needs to be serialized
-						// to BigInt or string.
-						coeffInteger: c.coeffInteger.toString(10),
-						coeffFrac: c.coeffFrac.toNumber(),
-						degree: c.degree.toNumber(),
-						negative: c.negative,
-					};
-				}
-			);
-
-			const parentParentHash: Hash = await this.getParentParentHash(
-				api,
-				parentHash,
-				block
-			);
-
-			const [version, multiplier] = await Promise.all([
-				api.rpc.state.getRuntimeVersion(parentParentHash),
-				api.query.transactionPayment.nextFeeMultiplier.at(parentHash),
-			]);
-
-			[specName, specVersion] = [
-				version.specName.toString(),
-				version.specVersion.toNumber(),
-			];
-
-			runtimeDoesNotMatch =
-				specName !== api.runtimeVersion.specName.toString() ||
-				specVersion !== api.runtimeVersion.specVersion.toNumber();
-
-			if (runtimeDoesNotMatch) {
-				const metadata = await api.rpc.state.getMetadata(parentParentHash);
-
-				decorated = expandMetadata(api.registry, metadata);
-			}
-
-			calcFee = CalcFee.from_params(
-				coefficients,
-				multiplier.toString(10),
-				perByte.toString(10),
+			// This particular runtime version is not supported with fee calcs or
+			// does not have the necessay materials to build calcFee
+			return {
+				specVersion,
 				specName,
-				specVersion
-			);
+			};
 		}
 
+		// Now that we know the exact runtime supports fee calcs, make sure we have
+		// the weights in the store
+		this.blockWeightStore[specVersion] ||= await this.getWeight(
+			api,
+			parentHash
+		);
+
+		const calcFee = CalcFee.from_params(
+			coefficients,
+			multiplier.toString(10),
+			perByte.toString(10),
+			specName,
+			specVersion
+		);
+
 		return {
 			calcFee,
 			specName,
 			specVersion,
-			decorated,
-			runtimeDoesNotMatch,
 		};
 	}
 
+	/**
+	 *	Get a formatted blockweight store value for the runtime corresponding to the given block hash.
+	 *
+	 * @param api ApiPromise
+	 * @param blockHash Hash of a block in the runtime to get the extrinsic base weight(s) for
+	 * @returns formatted block weight store entry
+	 */
+	private async getWeight(
+		api: ApiPromise,
+		blockHash: BlockHash
+	): Promise<WeightValue> {
+		const metadata = await api.rpc.state.getMetadata(blockHash);
+		const {
+			consts: { system },
+		} = expandMetadata(api.registry, metadata);
```

Comment:
We should keep an eye on what registry we use here in the future. There is a chance it may be better to pull the registry off `block` we get from the rpc response since that should be guranteed to have the exact types for the runtime. For now I _think_ this is fine.